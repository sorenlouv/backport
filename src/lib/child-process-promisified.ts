import childProcess from 'child_process';
import { promisify } from 'util';
import { logger } from './logger';
const execPromisified = promisify(childProcess.exec);

export async function exec(
  cmd: string,
  options: childProcess.ExecOptions & { cwd: string }
) {
  const res = await execPromisified(cmd, {
    maxBuffer: 100 * 1024 * 1024,
    ...options,

    // ensure that git commands return english error messages
    env: { ...process.env, LANG: 'en_US' },
  });

  return res;
}

export async function spawnPromise(
  cmd: string,
  cmdArgs: string[],
  cwd: string
): Promise<{
  cmdArgs: string[];
  code: number | null;
  stderr: string;
  stdout: string;
}> {
  const fullCmd = `${cmd} ${cmdArgs.join(' ')}`;
  logger.info(`Running command: "${fullCmd}"`);

  return new Promise(function (resolve, reject) {
    const subprocess = childProcess.spawn(cmd, cmdArgs, {
      cwd,

      // ensure that git commands return english error messages
      env: { ...process.env, LANG: 'en_US' },
    });
    let stderr = '';
    let stdout = '';

    subprocess.stdout.on('data', (data: string) => {
      stdout += data;
    });

    subprocess.stderr.on('data', (data: string) => {
      stderr += data;
    });

    subprocess.on('close', (code) => {
      if (code === 0 || code === null) {
        resolve({ cmdArgs, code, stderr, stdout });
      } else {
        const err = new SpawnError({ cmdArgs, code, stderr, stdout });
        logger.verbose(`Error when running command: "${fullCmd}"`, err);
        reject(err);
      }
    });

    subprocess.on('error', (err) => {
      reject(err);
    });
  });
}

export const spawnStream = (cmd: string, cmdArgs: ReadonlyArray<string>) => {
  return childProcess.spawn(cmd, cmdArgs, {
    env: { ...process.env, LANG: 'en_US' },
  });
};

export type SpawnErrorContext = {
  cmdArgs: string[];
  code: number;
  stderr: string;
  stdout: string;
};

export class SpawnError extends Error {
  context: SpawnErrorContext;
  constructor(context: SpawnErrorContext) {
    const cmdArgs = context.cmdArgs.join(' ');
    const message = `Code: ${
      context.code
    }, Args: "${cmdArgs}", Message: ${context.stderr.trim()}`;

    super(message);
    Error.captureStackTrace(this, SpawnError);
    this.name = 'SpawnError';
    this.message = message;
    this.context = context;
  }
}
