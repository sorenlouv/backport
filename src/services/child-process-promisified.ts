import child_process from 'child_process';
import { promisify } from 'util';
import { logger } from './logger';

export async function exec(cmd: string, options: child_process.ExecOptions) {
  logger.info(`exec cmd: ${cmd}`);
  const execPromisified = promisify(child_process.exec);
  const res = await execPromisified(cmd, {
    maxBuffer: 100 * 1024 * 1024,
    ...options,
  });
  logger.verbose(`exec result`, res);
  return res;
}

export const execAsCallback = (
  ...args: Parameters<typeof child_process.exec>
) => {
  return child_process.exec(...args);
};
