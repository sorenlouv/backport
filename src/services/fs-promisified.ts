import fs from 'fs';
import { promisify } from 'util';

export const writeFile = promisify(fs.writeFile);
export const readFile = promisify(fs.readFile);
export const stat = promisify(fs.stat);
export const chmod = promisify(fs.chmod);
