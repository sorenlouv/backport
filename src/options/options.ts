import { fetchDefaultRepoBranchAndPerformStartupChecks } from '../services/github/v4/fetchDefaultRepoBranchAndPerformStartupChecks';
import { PromiseReturnType } from '../types/PromiseReturnType';
import { setLogLevel, setRedactedAccessToken } from './../services/logger';
import { ConfigOptions } from './ConfigOptions';
import { getOptionsFromCliArgs } from './cliArgs';
import { getOptionsFromConfigFiles } from './config/config';
import { getValidatedOptions } from './getValidatedOptions';

export type BackportOptions = Readonly<PromiseReturnType<typeof getOptions>>;
export async function getOptions(
  argv: string[],
  optionsFromModule?: ConfigOptions
) {
  const optionsFromConfig = await getOptionsFromConfigFiles(
    optionsFromModule?.ci
  );
  const optionsFromCli = getOptionsFromCliArgs(
    { ...optionsFromConfig, ...optionsFromModule },
    argv
  );
  setRedactedAccessToken(optionsFromCli.accessToken);

  // set log level when all config options have been taken into account
  setLogLevel({ verbose: optionsFromCli.verbose });

  // TODO: move `getValidatedOptions` to `getOptionsFromCliArgs`
  const validatedOptions = getValidatedOptions(optionsFromCli);

  // TODO: make `username` optional by defaulting to `currentUsername`
  const { defaultBranch } = await fetchDefaultRepoBranchAndPerformStartupChecks(
    validatedOptions
  );

  return {
    ...validatedOptions,

    // use the default branch as source branch (normally master) unless an explicit `sourceBranch` has been given
    sourceBranch: validatedOptions.sourceBranch
      ? validatedOptions.sourceBranch
      : defaultBranch,
  };
}
