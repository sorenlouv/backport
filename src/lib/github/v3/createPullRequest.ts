import { Octokit } from '@octokit/rest';
import apm from 'elastic-apm-node';
import Handlebars from 'handlebars';
import { ora } from '../../../lib/ora';
import { ValidConfigOptions } from '../../../options/options';
import { PACKAGE_VERSION } from '../../../utils/packageVersion';
import { BackportError } from '../../BackportError';
import { getSourceBranchFromCommits } from '../../getSourceBranchFromCommits';
import { logger } from '../../logger';
import { Commit } from '../../sourceCommit/parseSourceCommit';
import { getFirstLine, getShortSha } from '../commitFormatters';
import { fetchExistingPullRequest } from '../v4/fetchExistingPullRequest';
import { getGithubV3ErrorMessage } from './getGithubV3ErrorMessage';

export interface PullRequestPayload {
  owner: string;
  repo: string;
  title: string;
  body: string;
  head: string;
  base: string;
  [key: string]: unknown;
}

export async function createPullRequest({
  options,
  prPayload,
}: {
  options: ValidConfigOptions;
  prPayload: PullRequestPayload;
}): Promise<{
  url: string;
  number: number;
  didUpdate: boolean;
}> {
  logger.info(
    `Creating PR with title: "${prPayload.title}". ${prPayload.head} -> ${prPayload.base}`,
  );

  const { accessToken, githubApiBaseUrlV3 } = options;
  const spinner = ora(options.interactive, `Creating pull request`).start();

  if (options.dryRun) {
    spinner.succeed();
    return { didUpdate: false, number: 1337, url: 'this-is-a-dry-run' };
  }

  const span = apm.startSpan('REST: Create pull request');

  try {
    const octokit = new Octokit({
      auth: accessToken,
      baseUrl: githubApiBaseUrlV3,
      log: logger,
    });

    const res = await octokit.pulls.create(prPayload);

    spinner.succeed();

    return {
      url: res.data.html_url,
      number: res.data.number,
      didUpdate: false,
    };
  } catch (e) {
    // retrieve url for existing
    try {
      const existingPR = await fetchExistingPullRequest({
        options,
        prPayload,
      });

      if (existingPR) {
        spinner.succeed('Updating existing pull request');
        return {
          url: existingPR.url,
          number: existingPR.number,
          didUpdate: true,
        };
      }
    } catch (e) {
      logger.error('Could not retrieve existing pull request', e);
      // swallow error
    } finally {
      span?.end();
    }

    spinner.fail();
    throw new BackportError(
      //@ts-expect-error
      `Could not create pull request: ${getGithubV3ErrorMessage(e)}`,
    );
  }
}

export function getPullRequestBody({
  options,
  commits,
  targetBranch,
}: {
  options: ValidConfigOptions;
  commits: Commit[];
  targetBranch: string;
}) {
  const commitMessages = commits
    .map((c) => {
      const message = c.sourcePullRequest
        ? `[${getFirstLine(c.sourceCommit.message)}](${
            c.sourcePullRequest.url
          })`
        : `${getFirstLine(c.sourceCommit.message)} (${getShortSha(
            c.sourceCommit.sha,
          )})`;

      return ` - ${message}`;
    })
    .join('\n');

  const sourceBranch = getSourceBranchFromCommits(commits);

  const defaultPrDescription =
    '# Backport\n\n' +
    'This will backport the following commits from `{{sourceBranch}}` to `{{targetBranch}}`:\n' +
    '{{commitMessages}}\n\n' +
    '<!--- Backport version: {{PACKAGE_VERSION}} -->\n\n' +
    '### Questions ?\n' +
    'Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)';

  const customPrDescription = options.prDescription
    ?.replaceAll('{{defaultPrDescription}}', defaultPrDescription)
    ?.replaceAll('{defaultPrDescription}', defaultPrDescription) // retain backwards compatibility
    ?.replaceAll('{{commitsStringified}}', JSON.stringify(commits))
    ?.replaceAll('{commits}', JSON.stringify(commits)); // retain backwards compatibility

  const template = Handlebars.compile(
    customPrDescription ?? defaultPrDescription,
  );

  return template({
    sourcePullRequest: commits[0].sourcePullRequest, // assume that all commits are from the same PR
    sourceBranch,
    targetBranch,
    commitMessages,
    PACKAGE_VERSION,
    commits,
  });
}

export function getTitle({
  options,
  commits,
  targetBranch,
}: {
  options: ValidConfigOptions;
  commits: Commit[];
  targetBranch: string;
}) {
  const sourceBranch = getSourceBranchFromCommits(commits);
  const commitMessages = commits
    .map((c) => getFirstLine(c.sourceCommit.message))
    .join(' | ');

  const defaultPrTitle = '[{{targetBranch}}] {{commitMessages}}';

  const template = Handlebars.compile(options.prTitle ?? defaultPrTitle, {
    noEscape: true,
  });

  return template({
    sourcePullRequest: commits[0].sourcePullRequest, // assume that all commits are from the same PR
    sourceBranch,
    targetBranch,
    commitMessages,
    commits,
  }).slice(0, 240);
}

Handlebars.registerHelper('shortSha', getShortSha);
