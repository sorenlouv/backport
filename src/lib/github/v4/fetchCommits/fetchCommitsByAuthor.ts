import { isEmpty, uniqBy, orderBy, first } from 'lodash';
import { graphql } from '../../../../graphql/generated';
import { CommitsByAuthorQuery } from '../../../../graphql/generated/graphql';
import { ValidConfigOptions } from '../../../../options/options';
import { filterNil } from '../../../../utils/filterEmpty';
import { filterUnmergedCommits } from '../../../../utils/filterUnmergedCommits';
import { BackportError } from '../../../BackportError';
import { swallowMissingConfigFileException } from '../../../remoteConfig';
import {
  Commit,
  parseSourceCommit,
} from '../../../sourceCommit/parseSourceCommit';
import { fetchAuthorId } from '../fetchAuthorId';
import { getGraphQLClient, GithubV4Exception } from './graphqlClient';

async function fetchByCommitPath({
  options,
  authorId,
  commitPath,
}: {
  options: {
    accessToken: string;
    dateSince: string | null;
    dateUntil: string | null;
    githubApiBaseUrlV4?: string;
    maxNumber?: number;
    repoName: string;
    repoOwner: string;
    sourceBranch: string;
  };
  authorId: string | null | undefined;
  commitPath: string | null;
}) {
  const {
    accessToken,
    dateSince,
    dateUntil,
    githubApiBaseUrlV4 = 'https://api.github.com/graphql',
    maxNumber = 10,
    repoName,
    repoOwner,
    sourceBranch,
  } = options;

  const query = graphql(`
    query CommitsByAuthor(
      $authorId: ID
      $commitPath: String
      $dateSince: GitTimestamp
      $dateUntil: GitTimestamp
      $maxNumber: Int!
      $repoName: String!
      $repoOwner: String!
      $sourceBranch: String!
    ) {
      repository(owner: $repoOwner, name: $repoName) {
        ref(qualifiedName: $sourceBranch) {
          target {
            ... on Commit {
              __typename
              history(
                first: $maxNumber
                author: { id: $authorId }
                path: $commitPath
                since: $dateSince
                until: $dateUntil
              ) {
                edges {
                  node {
                    __typename
                    ...SourceCommitWithTargetPullRequestFragment
                  }
                }
              }
            }
          }
        }
      }
    }
  `);

  const variables = {
    repoOwner,
    repoName,
    sourceBranch,
    maxNumber,
    authorId,
    commitPath,
    dateSince,
    dateUntil,
  };

  try {
    const client = getGraphQLClient({ accessToken, githubApiBaseUrlV4 });
    const result = await client.query(query, variables);

    if (result.error) {
      throw new GithubV4Exception(result);
    }

    return result.data;
  } catch (e) {
    if (e instanceof GithubV4Exception) {
      if (e.result.statusCode === 502 && maxNumber > 50) {
        throw new BackportError(
          `The GitHub API returned a 502 error. Try reducing the number of commits to display: "--max-number 20"`,
        );
      }
    }
    return swallowMissingConfigFileException<CommitsByAuthorQuery>(e);
  }
}

export async function fetchCommitsByAuthor(options: {
  accessToken: string;
  author: string | null;
  branchLabelMapping?: ValidConfigOptions['branchLabelMapping'];
  commitPaths?: string[];
  dateSince: string | null;
  dateUntil: string | null;
  githubApiBaseUrlV4?: string;
  maxNumber?: number;
  onlyMissing?: boolean;
  repoName: string;
  repoOwner: string;
  sourceBranch: string;
}): Promise<Commit[]> {
  const { sourceBranch, commitPaths = [] } = options;

  const authorId = await fetchAuthorId(options);
  const responses = (
    await Promise.all(
      isEmpty(commitPaths)
        ? [fetchByCommitPath({ options, authorId, commitPath: null })]
        : commitPaths.map((commitPath) =>
            fetchByCommitPath({ options, authorId, commitPath }),
          ),
    )
  ).filter(filterNil);

  // we only need to check if the first item is `null` (if the first is `null` they all are)
  if (first(responses)?.repository?.ref === null) {
    throw new BackportError(
      `The upstream branch "${sourceBranch}" does not exist. Try specifying a different branch with "--source-branch <your-branch>"`,
    );
  }

  const commits = responses
    .flatMap((res) => {
      const repoRefTarget = res.repository?.ref?.target;
      if (repoRefTarget?.__typename !== 'Commit') {
        return;
      }

      return repoRefTarget.history.edges?.map((edge) => {
        const sourceCommit = edge?.node;
        if (sourceCommit) {
          return parseSourceCommit({ options, sourceCommit });
        }
      });
    })
    .filter(filterNil);

  // terminate if not commits were found
  if (isEmpty(commits)) {
    const pathText =
      commitPaths.length > 0 ? ` touching files in path: "${commitPaths}"` : '';

    const errorText = options.author
      ? `There are no commits by "${options.author}" in this repository${pathText}. Try with \`--all\` for commits by all users or \`--author=<username>\` for commits from a specific user`
      : `There are no commits in this repository${pathText}`;

    throw new BackportError(errorText);
  }

  const commitsUnique = uniqBy(commits, (c) => c.sourceCommit.sha);
  const commitsSorted = orderBy(
    commitsUnique,
    (c) => c.sourceCommit.committedDate,
    'desc',
  );

  if (options.onlyMissing) {
    return commitsSorted.filter(filterUnmergedCommits);
  }

  return commitsSorted;
}
