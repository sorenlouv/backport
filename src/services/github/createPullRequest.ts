import axios, { AxiosResponse } from 'axios';
import { BackportOptions } from '../../options/options';
import { GithubIssue } from './GithubApiTypes';
import { handleGithubError } from './handleGithubError';
import { logger } from '../logger';

export async function createPullRequest(
  { apiHostname, repoName, repoOwner, accessToken }: BackportOptions,
  payload: {
    title: string;
    body: string;
    head: string;
    base: string;
  }
) {
  logger.info(
    `Creating PR with title: "${payload.title}". ${payload.head} -> ${payload.base}`
  );

  try {
    const res: AxiosResponse<GithubIssue> = await axios.post(
      `https://${apiHostname}/repos/${repoOwner}/${repoName}/pulls?access_token=${accessToken}`,
      payload
    );
    return {
      html_url: res.data.html_url,
      number: res.data.number
    };
  } catch (e) {
    throw handleGithubError(e);
  }
}
