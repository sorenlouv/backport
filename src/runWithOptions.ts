import { BackportOptions } from './options/options';
import { verifyAccessToken } from './services/github/verifyAccessToken';
import { doBackportVersions } from './ui/doBackportVersions';
import { getBranches } from './ui/getBranches';
import { getCommits } from './ui/getCommits';
import { maybeSetupRepo } from './ui/maybeSetupRepo';
import { logger } from './services/logger';

export async function runWithOptions(options: BackportOptions) {
  await verifyAccessToken(options);

  const commits = await getCommits(options);
  const branches = await getBranches(options);

  await maybeSetupRepo(options);

  logger.info(`Backporting ${JSON.stringify(commits)} ${branches.join(',')}`);
  await doBackportVersions(options, commits, branches);
}
