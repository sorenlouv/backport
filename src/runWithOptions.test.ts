import axios from 'axios';
import inquirer from 'inquirer';
import { BackportOptions } from './options/options';
import { runWithOptions } from './runWithOptions';
import * as childProcess from './services/child-process-promisified';
import * as fs from './services/fs-promisified';
import * as createPullRequest from './services/github/v3/createPullRequest';
import * as fetchCommitsByAuthor from './services/github/v4/fetchCommitsByAuthor';
import { commitsWithPullRequestsMock } from './services/github/v4/mocks/commitsByAuthorMock';
import { SpyHelper } from './types/SpyHelper';

describe('runWithOptions', () => {
  let rpcExecMock: SpyHelper<typeof childProcess.exec>;
  let rpcExecOriginalMock: SpyHelper<typeof childProcess.execAsCallback>;
  let inquirerPromptMock: SpyHelper<typeof inquirer.prompt>;
  let axiosRequestSpy: SpyHelper<typeof axios.request>;

  afterEach(() => {
    jest.clearAllMocks();
  });

  beforeEach(async () => {
    const options: BackportOptions = {
      accessToken: 'myAccessToken',
      all: false,
      assignees: [],
      author: 'sqren',
      autoFixConflicts: undefined,
      branchLabelMapping: undefined,
      dryRun: false,
      editor: 'code',
      fork: true,
      gitHostname: 'github.com',
      githubApiBaseUrlV3: 'https://api.github.com',
      githubApiBaseUrlV4: 'https://api.github.com/graphql',
      mainline: undefined,
      maxNumber: 10,
      multipleBranches: false,
      multipleCommits: false,
      noVerify: true,
      path: undefined,
      prDescription: 'myPrDescription',
      prTitle: 'myPrTitle {targetBranch} {commitMessages}',
      pullNumber: undefined,
      repoName: 'kibana',
      repoOwner: 'elastic',
      resetAuthor: false,
      sha: undefined,
      sourceBranch: 'mySourceBranch',
      sourcePRLabels: [],
      prFilter: undefined,
      targetBranches: [],
      targetBranchChoices: [
        { name: '6.x' },
        { name: '6.0' },
        { name: '5.6' },
        { name: '5.5' },
        { name: '5.4' },
      ],
      targetPRLabels: [],
      username: 'sqren',
      verbose: false,
    };

    rpcExecMock = jest
      .spyOn(childProcess, 'exec')
      .mockResolvedValue({ stdout: 'success', stderr: '' });
    rpcExecOriginalMock = jest.spyOn(childProcess, 'execAsCallback');

    jest.spyOn(fs, 'writeFile').mockResolvedValue(undefined);

    jest.spyOn(fetchCommitsByAuthor, 'fetchCommitsByAuthor');
    jest.spyOn(createPullRequest, 'createPullRequest');

    inquirerPromptMock = jest
      .spyOn(inquirer, 'prompt')
      .mockImplementationOnce((async (args: any) => {
        return { promptResult: args[0].choices[0].value };
      }) as any)
      .mockImplementationOnce((async (args: any) => {
        return { promptResult: args[0].choices[0].name };
      }) as any);

    // Mock Github v4 API
    jest
      .spyOn(axios, 'post')

      // mock author id
      .mockResolvedValueOnce({
        data: {
          data: {
            user: {
              id: 'sqren_author_id',
            },
          },
        },
      })

      // mock list of commits
      .mockResolvedValueOnce({
        data: {
          data: commitsWithPullRequestsMock,
        },
      });

    // Mock Github v3 API
    axiosRequestSpy = jest
      .spyOn(axios, 'request')

      // mock create pull request
      .mockResolvedValueOnce({
        data: {
          html_url: 'pull request url',
          number: 1337,
        },
      });

    await runWithOptions(options);
  });

  it('getCommit should be called with correct args', () => {
    expect(fetchCommitsByAuthor.fetchCommitsByAuthor).toHaveBeenCalledWith(
      expect.objectContaining({
        repoName: 'kibana',
        repoOwner: 'elastic',
        username: 'sqren',
        githubApiBaseUrlV4: 'https://api.github.com/graphql',
      })
    );
  });

  it('createPullRequest should be called with correct args', () => {
    expect(createPullRequest.createPullRequest).toHaveBeenCalledWith({
      options: expect.objectContaining({
        repoName: 'kibana',
        repoOwner: 'elastic',
        githubApiBaseUrlV4: 'https://api.github.com/graphql',
      }),
      commits: [
        {
          existingTargetPullRequests: [],
          formattedMessage: 'Add ðŸ‘» (2e63475c)',
          pullNumber: undefined,
          selectedTargetBranches: [],
          sha: '2e63475c483f7844b0f2833bc57fdee32095bacb',
          sourceBranch: 'mySourceBranch',
        },
      ],
      targetBranch: '6.x',
      backportBranch: 'backport/6.x/commit-2e63475c',
    });
  });

  it('should make correct request when creating pull request', () => {
    expect(axiosRequestSpy).toHaveBeenCalledWith({
      auth: { password: 'myAccessToken', username: 'sqren' },
      data: {
        base: '6.x',
        body: `Backports the following commits to 6.x:\n - Add ðŸ‘» (2e63475c)\n\nmyPrDescription`,
        head: 'sqren:backport/6.x/commit-2e63475c',
        title: 'myPrTitle 6.x Add ðŸ‘» (2e63475c)',
      },
      method: 'post',
      url: 'https://api.github.com/repos/elastic/kibana/pulls',
    });
  });

  it('prompt calls should match snapshot', () => {
    expect(inquirer.prompt).toHaveBeenCalledTimes(2);
    expect(inquirerPromptMock.mock.calls).toMatchSnapshot();
  });

  it('exec should be called with correct args', () => {
    expect(rpcExecMock).toHaveBeenCalledTimes(10);
    expect(rpcExecMock.mock.calls).toMatchSnapshot();
  });

  it('git clone should be called with correct args', () => {
    expect(rpcExecOriginalMock).toHaveBeenCalledTimes(1);
    expect(rpcExecOriginalMock.mock.calls).toMatchSnapshot();
  });
});
