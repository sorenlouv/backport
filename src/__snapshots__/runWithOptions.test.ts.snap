// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runWithOptions exec should be called with correct args 1`] = `
Array [
  Array [
    "git remote rm origin",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git remote rm sqren",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git remote add sqren https://myAccessToken@github.com/sqren/kibana.git",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git remote rm elastic",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git remote add elastic https://myAccessToken@github.com/elastic/kibana.git",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git reset --hard && git clean -d --force && git fetch elastic 6.x && git checkout -B backport/6.x/commit-2e63475c elastic/6.x --no-track",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git fetch elastic master:master --force && git cherry-pick 2e63475c483f7844b0f2833bc57fdee32095bacb",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git push sqren backport/6.x/commit-2e63475c:backport/6.x/commit-2e63475c --force",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git checkout master && git branch -D backport/6.x/commit-2e63475c",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
]
`;

exports[`runWithOptions git clone should be called with correct args 1`] = `
Array [
  Array [
    "git clone https://myAccessToken@github.com/elastic/kibana.git --progress",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
]
`;

exports[`runWithOptions prompt calls should match snapshot 1`] = `
Array [
  Array [
    Array [
      Object {
        "choices": Array [
          Object {
            "name": "[90m1.[39m Add ðŸ‘» (2e63475c) ",
            "short": "Add ðŸ‘» (2e63475c)",
            "value": Object {
              "existingBackports": Array [],
              "message": "Add ðŸ‘» (2e63475c)",
              "pullNumber": undefined,
              "sha": "2e63475c483f7844b0f2833bc57fdee32095bacb",
            },
          },
          Object {
            "name": "[90m2.[39m Add witch (#85) ",
            "short": "Add witch (#85)",
            "value": Object {
              "existingBackports": Array [],
              "message": "Add witch (#85)",
              "pullNumber": 85,
              "sha": "f3b618b9421fdecdb36862f907afbdd6344b361d",
            },
          },
          Object {
            "name": "[90m3.[39m Add SF mention (#80) [32m6.3[39m",
            "short": "Add SF mention (#80)",
            "value": Object {
              "existingBackports": Array [
                Object {
                  "branch": "6.3",
                  "state": "MERGED",
                },
              ],
              "message": "Add SF mention (#80)",
              "pullNumber": 80,
              "sha": "79cf18453ec32a4677009dcbab1c9c8c73fc14fe",
            },
          },
          Object {
            "name": "[90m4.[39m Add backport config (3827bbba) ",
            "short": "Add backport config (3827bbba)",
            "value": Object {
              "existingBackports": Array [],
              "message": "Add backport config (3827bbba)",
              "pullNumber": undefined,
              "sha": "3827bbbaf39914eda4f02f6940189844375fd097",
            },
          },
          Object {
            "name": "[90m5.[39m Initial commit (5ea0da55) ",
            "short": "Initial commit (5ea0da55)",
            "value": Object {
              "existingBackports": Array [],
              "message": "Initial commit (5ea0da55)",
              "pullNumber": undefined,
              "sha": "5ea0da550ac191029459289d67f99ad7d310812b",
            },
          },
        ],
        "message": "Select commit to backport",
        "name": "promptResult",
        "pageSize": 5,
        "type": "list",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "choices": Array [
          Object {
            "name": "6.x",
          },
          Object {
            "name": "6.0",
          },
          Object {
            "name": "5.6",
          },
          Object {
            "name": "5.5",
          },
          Object {
            "name": "5.4",
          },
        ],
        "message": "Select branch to backport to",
        "name": "promptResult",
        "type": "list",
      },
    ],
  ],
]
`;
